### DISCUSSION TITLE
Updating Buffer During Playback

### DISCUSSION THREAD

**Discussion by rozgonyiadam** [naudio/NAudio#1105] [Q&A] [answered]
How to implement playback that takes audio data from a circular buffer (CircleBuffer) and periodically updates or expands it during playback?
```
readonly BufferedWaveProvider bufferedWaveProvider = new BufferedWaveProvider(new WaveFormat());

public void AddChunkToBuffer(MemoryStream chunk)
{
    if (chunk != null && chunk.Length > 0)
    {
        byte[] chunkData = chunk.ToArray();
        bufferedWaveProvider.AddSamples(chunkData, 0, chunkData.Length);
    }
}

// ...

audioPlayer.Init(bufferedWaveProvider);
AddChunkToBuffer(stream);
```

The playback is operational; however, I am uncertain about **how to append the next chunk**. Considering there is a length limit for `BufferedWaveProvider`, it throw a `buffer full` exception.
If I need to delete the already played portion of the buffer (from the beginning) before appending, how can I achieve this?

**Comment by markheath** [comment, marked as answer]
You can set the `BufferDuration` property on a `BufferedWaveProvider` before reading from it. The default is 5 seconds, but you could make it longer. There is also a `ClearBuffer` method you can use to clear it. There is also a `DiscardOnBufferOverflow` you can set to true if you don't want buffer full exceptions

**Reply by rozgonyiadam** [reply]
Thank you Mark! I fine-tuned the buffer size with the `BufferDuration` property, so `bufferedWaveProvider` can now accommodate the next chunk without the 'buffer full' exception.

There's only one thing hindering the continuous update of the buffer now: playback pauses for a few seconds and then resumes when it reaches the newly added data in buffer. Naturally, the larger the audio chunk size, the fewer 'gaps' are experienced. Do you have any ideas on why playback is not continuous?

```
IDispatcherTimer timer = Application.Current.Dispatcher.CreateTimer();
timer.Interval = TimeSpan.FromSeconds(1); //TimeSpan.FromSeconds(1);
timer.Tick += OnTimeChanging;

private void OnTimeChanging(object sender, EventArgs e)
{
   AddChunkToBuffer(newStream);
}
```

**Reply by markheath** [reply]
that usually means you are playing back audio faster than you are adding it to the buffer.

**Reply by rozgonyiadam** [reply]
Thanks! The issue was actually caused by improper thread handling - precisely the lack thereof. I am testing the expandability of the buffer, and then I will close the issue as 'Resolved'.
